src/game_battler.cpp

```diff
@@ -77,13 +78,18 @@ RPG::State::Restriction Game_Battler::GetSignificantRestriction() const {
 	return State::GetSignificantRestriction(GetStates());
 }
 
-bool Game_Battler::CanAct() const {
-	const std::vector<int16_t> states = GetInflictedStates();
-	for (int i = 0; i < (int)states.size(); i++) {
-		// States are guaranteed to be valid
-		const RPG::State* state = ReaderUtil::GetElement(Data::states, states[i]);
-		if (state->restriction == RPG::State::Restriction_do_nothing) {
-			return false;
+bool Game_Battler::CanActImpl(int max_arp) const {
+	const auto& states = GetStates();
+	for (size_t i = 0; i < states.size(); ++i) {
+		if (states[i] > 0) {
+			const auto* state = ReaderUtil::GetElement(Data::states, i + 1);
+			if (state
+					&& state->restriction == RPG::State::Restriction_do_nothing
+					&& state->auto_release_prob <= max_arp
```

Comment:
Imo ``CanActOrRecoverable`` in confusing because it does two completely different things in one function. In one case it decides about ``Can Act right now`` and in the other case it decides about ``Can currently not act, but has a recoverable state``.

So these should be two distinct functions. Passing in 0 or max to get two completely different things makes this hard to understand (I had to think for 10 minutes about why this func works).

Reply:
Is it not better to have one function for this instead of having to reproduce something like `CanAct() || !HasUnrecoverableState()` in all places this is used?

Grepping for this in the code:
```
game_battle.cpp
162:            if (!actor->IsHidden() && actor->CanActOrRecoverable()) {

game_battler.h
126:    bool CanActOrRecoverable() const;
898:inline bool Game_Battler::CanActOrRecoverable() const {

game_battler.cpp
662:    // the gauge and carry it between battles if !CanActOrRecoverable().

scene_battle_rpg2k3.cpp
55:     if (battler.IsHidden() || !battler.CanActOrRecoverable()) {
```

And it's used another time in the placement PR.

The private `CanActImpl()` method is a bit hard to read. I did it that way to avoid copy and paste, but I can split these 2 methods out if that makes it easier.